### PR TITLE
Store sensitive configuration as encrypted data

### LINKED ISSUE #29676: Store sensitive configuration as encrypted data
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md)
- [X] I have already [searched in existing features request](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3AImprovement%2CFeature%2CDeveloper+Feature+) and found no previous suggestion of this feature.

### Is your feature request related to a problem?

Some sensitive data like API key of payment, delivery, ... or native SMTP password are stored as a clear plain text in the database.

In case of SQL Injection exploit, a hacker can in chain access and update merchant values by corrupted data to divert a stream of payments, emails, employee accounts, ...

### Describe the solution you'd like

Configurations of PrestaShop are today very flexible and interesting to manage module configuration by language, shops, ...
It could be nice to store sensitive configuration as encrypted data.

### Alternatives you've considered

_No response_

### Additional context

I make a Pull Request as a POC.

### Do you plan to work on this feature?

- [ ] I'm willing to contribute a [formal specification](https://github.com/PrestaShop/prestashop-specs).
- [ ] I'm willing to provide any wireframes or design assets required for this feature.
- [X] I'm willing to submit a Pull Request that implements this feature.
- [X] I'm willing to help verify that the implemented feature works as intended and produces no unintended side effects.

### PR DESCRIPTION
**Pull request by clotairer**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | This pull request is a POC of a solution to store encrypted configuration in database.<br>This is work in progress. I develop a CryptedConfiguration class and the adapter associated but wait for a feedback from PrestaShop Core team to follow and improve my idea.</br>At least, it could be nice in an other PR to encrypt native sentive data like SMTP email configuration.
| Type?             | improvement
| Category?         | CO
| BC breaks?        | no
| Deprecations?     |  no
| Fixed ticket?     | Fixes #29676
| How to test?      | This is cover by Unit test.<br>In a module, just use `CryptedConfiguration::updateValue('CONFIG_KEY', 'clear value');` to store a configuration and `CryptedConfiguration::get('CONFIG_KEY')` to get the clear value stored in the database.
| Possible impacts? | None.

### CODE REVIEW

**Comment by FabienPapet** [comment]
I think this is a good idea but it can't be merged as it is. IMO we should not create more legacy classes and create new core classes instead

**Comment by clotairer** [comment]
* First of all, could you validate the algotythm ? in term of security ? long therm maintenance of phpEncrytation class ? After creating lot of configuration, upgrade will be quite hard.
* OK about legacy class, please consider I will provide it as backward compatibility for module editor that need a scope compatibility from 7.1 until develop. I know that, as explain, I don't expected you merge this POC but lay a first brick to go on with PrestaShop core team advice.

File: classes/CryptedConfiguration.php [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+     * Get global value.
+     *
+     * @param string $key Configuration key
+     * @param int|null $idLang Language ID
+     *
+     * @return string
+     */
+    public static function getGlobalValue($key, $idLang = null)
+    {
+        return CryptedConfiguration::get($key, $idLang, 0, 0);
```
Comment (atomiix) [inline classes/CryptedConfiguration.php:80]:
```suggestion
        return static::get($key, $idLang, 0, 0);
```

File: classes/CryptedConfiguration.php [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+     *
+     * @return array Values for all shops
+     */
+    public static function getMultiShopValues($key, $idLang = null)
+    {
+        $shops = Shop::getShops(false, null, true);
+        $resultsArray = [];
+        foreach ($shops as $idShop) {
+            $resultsArray[$idShop] = CryptedConfiguration::get($key, $idLang, null, $idShop);
+        }
```
Comment (atomiix) [inline classes/CryptedConfiguration.php:97]: I think we should rely on the parent class to get the data and just add the decryption layer, WDYT?
```suggestion
        $resultsArray = parent::getMultiShopValues($key, $idLang);
        foreach ($resultsArray as $idShop => $result) {
            if (empty($result) {
                continue;
            }
            $resultsArray[$idShop] = self::getPhpEncryption()->decrypt($rawValue);
        }
```

File: classes/CryptedConfiguration.php [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+            $idShop = Shop::getContextShopID(true);
+        }
+        if ($idShopGroup === null) {
+            $idShopGroup = Shop::getContextShopGroupID(true);
+        }
+
+        $results = [];
+        foreach ($keys as $key) {
+            $results[$key] = CryptedConfiguration::get($key, $idLang, $idShopGroup, $idShop);
+        }
```
Comment (atomiix) [inline classes/CryptedConfiguration.php:131]: Same here:
```suggestion
        $results = parent::($keys, $idLang, $idShopGroup, $idShop);
        for ($results in $key => $result) {
            if (empty($result) {
                continue;
            }
            $results[$key] = self::getPhpEncryption()->decrypt($result);
        }
```

File: classes/CryptedConfiguration.php [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+     *
+     * @param string $key
+     * @param mixed $values
+     * @param bool $html
+     *
+     * @return bool
+     */
+    public static function updateGlobalValue($key, $values, $html = false)
+    {
+        return CryptedConfiguration::updateValue($key, $values, $html, 0, 0);
```
Comment (atomiix) [inline classes/CryptedConfiguration.php:168]:
```suggestion
        return static::updateValue($key, $values, $html, 0, 0);
```

**Review by atomiix — COMMENTED** [review]
For the class `CryptedConfiguration` I'm wondering if it would be worth it to change the constructor of `Adapter/Configuration` to pass which `class/CryptedConfiguration`/`class/Configuration` to use so then in your `CryptedConfiguration` you could simply pass `class/CryptedConfiguration` and every thing would work without having to override every method.

Other than that, relying on `PhpEncryption` seems pretty safe for me!

**Comment by matks** [comment]
@atomiix @clotairer so what's the next move?

By the way I agree with @atomiix about validating the algorythm

**Comment by clotairer** [comment]
@matks I worked on a new version this week by removing the CryptedConfiguration classes witch is a legacy class. I update the cryted configuration adapter to use his parent. And I build a sample demo module using services. I must create new unit tests…

Moreover I’d like to create a polyfill to incluse classes with composer on modules.

ETA to deliver a new commit: this weekend.

**Comment by clotairer** [comment]
I’ve an issue with the algorythm. We need to add a salt for each configuration. If you use the same hash like the cookie key between two configuration you will obtain the same encrypted value. And that’s big issue regarding the security.

**Comment by matks** [comment]
> I’ve an issue with the algorythm. We need to add a salt for each configuration. If you use the same hash like the cookie key between two configuration you will obtain the same encrypted value. And that’s big issue regarding the security.

It can be issue that value1 and value2 have the encrypted value if attacker knows value2. Then he can guess value1 as the 2 encrypted values are identical.
You could add a unique salt to each row? So each each value is encrypted using the cookie key salt + the unique row salt.

**Comment by clotairer** [comment]
Yes, we can store as configuration value the salt separated by a characters like a $ for instance + the encrypted value.

**Comment by atomiix** [comment]
We can maybe use the configuration name as the salt?

**Review by jolelievre — COMMENTED** [review]
Hi @clotairer 

I love this idea, but I wonder if this should be handled by a dedicated CryptedConfiguration class, it could be handled in the regular Configuration class, but it would imply adding a column in the configuration table to indicate if the value is secured or not

Then depending on this option the Configuration class would automatically perform the encryption in get/set The drawback is that it requires an update in the DB and it also means it wouldn't be possible to propose a polyfill for older versions as you suggested since they won't have this extra column

Or another alternative could be to propose a mix a both solutions:
- in v9 we add this extra column and the Configuration automatically handles the encryption part
- in older version we propose a polyfill with the explicit `CryptedConfiguration` so it can still be used without the extra column but the module must explicitly use it This use case would become overkill in v9 if you activate the field option in DB

**Comment by Hlavtox** [comment]
Please go for EncryptedConfiguration, it will be more understandable :-)

**Comment by clotairer** [comment]
You are right as well.
I try to do something one year ago as a POC not as a real solution.

I sould closed this obsolete PR and star a new one ?

In fact my idea was to create a Polyfill to backport the feature in a module for a larger compatibility. There is no cases in PrestaShop, but I think it can open new ways to improve core and clean legacy. Perhaps we should talk about a technical specification about this on #32238

thanks

**Pull request closed without merging**